src/mahalanobis.jl

```diff
@@ -1,19 +1,26 @@
 # Mahalanobis distances
 
-struct Mahalanobis{T} <: Metric
-    qmat::Matrix{T}
+struct Mahalanobis{M<:AbstractMatrix} <: Metric
+    qmat::M
 end
 
-struct SqMahalanobis{T} <: SemiMetric
-    qmat::Matrix{T}
+struct SqMahalanobis{M<:AbstractMatrix} <: SemiMetric
+    qmat::M
 end
 
-result_type(::Mahalanobis{T}, ::Type, ::Type) where {T} = T
-result_type(::SqMahalanobis{T}, ::Type, ::Type) where {T} = T
+function result_type(d::Mahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
+    z = oneunit(T1) - oneunit(T2)
+    return typeof(z * oneunit(eltype(d.qmat)) * z)
+end
```

Comment:
IMO this is not completely correct, and should be changed to
```suggestion
function result_type(d::Mahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
    z = oneunit(T1) - oneunit(T2)
    return typeof(sqrt(z * oneunit(eltype(d.qmat)) * z))
end
```
or
```suggestion
function result_type(d::Mahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
    z = zero(T1) - zero(T2)
    return typeof(sqrt(z * zero(eltype(d.qmat)) * z))
end
```
(I noticed that maybe it's more natural to use the additive identity element instead of `oneunit` here)
It could also just be based on the `result_type` for `SqMahalanobis` by writing
```suggestion
function result_type(d::Mahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
    return typeof(sqrt(zero(result_type(SqMahalanobis(d.qmat), T1, T2))))
end
```